### ISSUE TITLE
Accelerate variables lookup

### ISSUE THREAD

**Issue by mougino** [RFO-BASIC/Basic#198]
While discussing globals [on the forum] (http://rfobasic.freeforums.org/global-variables-was-multiple-values-for-bundle-get-t3760.html) user Gikam asked for [binary search on variables] (http://rfobasic.freeforums.org/post24358.html#p24358) instead of current linear search.

Marc, you already investigated keywords lookup, but not variable lookups, and according to you back in 2012, variable lookup is the 1st most consuming task of the parser.

I investigated [here] (http://rfobasic.freeforums.org/post24370.html#p24370) and just built a BASIC! test version perfecting this trail. I wrote a benchmark reading randomly 500 out of 702 different variables: with BASIC! v01.88 it takes an average of 250ms. On my modified BASIC! (based on V01.88) it takes an average of 52ms (gain of 80%).

Changes only take place in 2 functions: `Run.searchVar(String)` and `Run.createNewVar(String, int)` as indicated [here] (http://rfobasic.freeforums.org/post24370.html#p24370)

**Comment by mougino** [comment]
Out of curiosity I ran sample program `f22_benchmark.bas`: on regular BASIC! average on 10 passes is 2.12s ; on modified BASIC! with var binary it is 0.2s

**Comment by mougino** [comment]
Whole modifications to `Run.java`:
```Java
	private boolean searchVar(String name) {		// search for a variable by name
		int j = Collections.binarySearch(
            VarNames.subList(VarSearchStart, VarNames.size()),	// VarSearchStart is usually zero but will change when executing User Function
            name);
        if (j < 0) {								// not in list of variable names
          VarIsNew = true;
          return false;
        } else {								      // found it
          if (VarIsArray) {
            ArrayDescriptor array = ArrayTable.get(VarIndex.get(j));
            if (!array.valid()) {			// array invalidated through a different variable
              VarNames.set(j, " ");		// clear this variable so a new one with the same name can be created
              VarIsNew = true;
              return false;
            }
          }
          VarIsNew = false;
          VarNumber = j;
          theValueIndex = VarIndex.get(j);	// get the value index from the var table
          return true;
        }
	}
```
And
```Java
	private int createNewVar(String name, int val) {// make a new var table entry; val is an index into one of the lists
        int index = Collections.binarySearch(
                VarNames.subList(VarSearchStart, VarNames.size()),	// VarSearchStart is usually zero but will change when executing User Function
                name);                        // index returned by binarySearch() is the place where to insert var
        index = VarSearchStart - index - 1;   // so that VarNames.subList() is sorted alphabetically
        VarNames.add(index, name);						// create entry in list of variable names
        VarIndex.add(index, val);							// create corresponding index list entry
        VarNumber++;
        return index;
    }
```
Improve BASIC! programs by a speed factor x5 up to x10.

Although I have a crash of BASIC! when running my [simple SetG() / GetG() program] (http://rfobasic.freeforums.org/post24287.html#p24287).
So I guess my modifications to `CreateNewVar()` present a flaw... Unfortunately I don't have any simulator/debugger to find where.
If we fix it we could really improve BASIC! speed spectacularly :racehorse:

**Comment by jMarcS** [comment]
> Marc, you already investigated keywords lookup, but not variable lookups,

I'm reasonably sure that I investigated variable lookups, too, but that exercise was lost in the crunch at the end of last year when I took a couple of months mostly off of BASIC!. It would take some time to reconstruct what I did then. Instead, let's pursue your `binarySearch()` change.

> and according to you back in 2012, variable lookup is the 1st most consuming task of the parser.

That's right. I did not expect to see the kind of performance you are seeing, though!

You have translated `searchVar()` correctly, as far as I can tell. The problem is that you are only sorting the variable name list, `VarNames`. In the forum I described how variables are stored. It's pretty complicated. Specifically, you are not sorting `VarIndex`. Your variable look-ups are almost all referencing the wrong data storage. Try a simple test that checks the values of the variables.

I made changes a while ago that make this simpler: there used to be a list of string values and a list of numeric values, and the index in `VarIndex` pointed into one of them (or something else). Now those lists are combined into `Vars`, a list of all scalar values.

`Vars` is a list of references to `Var` objects. Currently those are either `NumVar` or `StrVar`, which hold scalar values. I'd like to extend that idea so they can also be arrays or functions or whatever. Once that happens, the index in `VarIndex` can be pulled into the `Var` class. Not literally, but conceptually, as part of the subclass hierarchy. I described some of the consequences in the forum. The mind boggles.

So that's the beginning of my "Grand Unified Variable" architecture. It's nothing new, I'm probably copying Perl, at least in spirit.

However, you don't need all that to make variable look-ups faster. I'm so dazzled by the potential glories of the GUV that I forgot to look again at the much more immediate problem of variable search speed. "Le meglio è l'inimico del bene."

This post is too long. I'll start another.

**Comment by mougino** [comment]
I do sort `VarIndex` at the same time as `VarNames`:
```Java
        index = VarSearchStart - index - 1;   // so that VarNames.subList() is sorted alphabetically
        VarNames.add(index, name);                      // create entry in list of variable names
        VarIndex.add(index, val);                           // create corresponding index list entry
        VarNumber++;
```

**Comment by jMarcS** [comment]
Normally I would wait until other infrastructure changes were complete. But the performance numbers you report are amazing. That makes it worthwhile to look at in intermediate change, just for variable storage and look-up. I've already made some relevant changes in earlier releases, so this won't be as hard as if we had tried to do it a couple of years ago.

I believe that your scheme will work if you sort `VarIndex` into the same order as `VarNames`. There are two ways to do that:
1. Modify the sort so it sorts both arrays based on the order of one of them.
2. Modify the data structure so there is only one list.

I'll describe both methods, but I have a very strong bias toward the second one, the data-oriented approach. That's obvious, from my previous post!

1. The algorithmic approach:

If you were writing your own search, this would be easy. Just implement an indirect sort: it's a classic classroom exercise to sort an index list and apply it to a content list. Since you're using a library method, it's not so easy.

This is such a common problem that I would not be surprised if Java already has such a method. I have not looked. That would make the solution very simple indeed.

2. The data-oriented approach:

Make the two lists into one list. This is not the "Grand Unified Variable" class. You just need a simple class that holds a `String` (the variable name) and an `int` (the value index). The value index would point at some other list, just as a `VarIndex` value does now. The code that uses the index knows by its context which list to use (`Vars`, `ArrayTable`, or `FunctionTable`).

Provide simple setters and getters so the rest of the code does not rely on the internal structure of the new class. Some day we'll be merging this intermediate change with the GUV!

Then go through ALL of the code and change every place that `VarNames` or `VarIndex` is used. This is not trivial, but like I said, I've made a lot of changes that were specifically intended to make this kind of change easier. To get an idea of the size of the problem, I deleted the declarations of both lists. That makes 45 compile errors. I should say "only" 45 compile errors -- not as bad as I expected.

**Comment by jMarcS** [comment]
Ha! I must laugh at myself. I just read what you wrote in the forum, and your new post here, and now I see I did not understand all of your earlier first post here. Yes, you do keep `VarIndex` in sync with `VarNames`. That's a nice variation on the "algorithmic approach".

I don't have a quick answer. I sense a long and tedious debugger session in my future!

**Comment by jMarcS** [comment]
VarNumber is not the number of variables. It is the index of the most-recently-searched variable. (One objective of my GUV plan is to eliminate more of those globals!) In `createNewVar(String, int)`:

```Java
        index = VarSearchStart - index - 1;   // so that VarNames.subList() is sorted alphabetically
        VarNames.add(index, name);                      // create entry in list of variable names
        VarIndex.add(index, val);                           // create corresponding index list entry
        VarNumber = index;       <-- not VarNumber++
        return index;
```

That fixes one crash, caused by an attempt to assign a numeric value to a string, or vice versa.

However, there is another crash, caused by an index out of bounds exception. In your test program, it happens in Bundle.Put inside a function. In my simpler test program, it's in an implicit LET (simple assignment) inside a function. If I don't do a function call, it does not crash.

Maybe just the first variable creation inside a function call?

**Comment by jMarcS** [comment]
Before your change, `createNewVar(String, int)` did not reference `VarSearchStart`. Now it does. But `VarSearchStart` is not set until the end of `doUserFunction()` -- too late to use it to create the local variables for the function arguments.

There's something tricky in there, indicated by the comment, "Now that all new variables have been created in main name space, start the function name space with the function parameter names."

This is from a bugfix. Suppose a user calls a function with an argument, like `fn(a)`, but `a` does not exist. A new `a` is created in the caller's name space.

So we can't create any variables in the called function's name space, or set the value of `VarSearchStart`, until all of the arguments have been processed. The bugfix is right where that comment is: there's a loop there that creates new variables. It does not call `createNewVar()`. It knows there are no variables in the function space yet, so it just does a fast, simple loop to stick the new variables at the end of the name space. They are not sorted.

This is an example of how optimizing for speed can hurt maintainabilty!

The fix is to set `VarSearchStart` before creating the local variables for the parameter arguments, and to call `createNewVar()` to create them. (I have not tried this!)

**Comment by jMarcS** [comment]
If `VarSearchStart` is set early, it is possible for either a RunTimeError("Too few calling parameters") or a syntax error (no closing parenthesis) to leave `VarSearchStart` with the wrong value. I think this is okay because those are not recoverable errors. A user could trap them with `OnError:`, maybe just to report the error string. There have been enough changes to globals that any attempt at recovery would (should!!) end badly. (It is unfortunate that we don't have a way of marking some errors unrecoverable.)

**Comment by mougino** [comment]
Wow :fearful: it's really like we pulled a bit of a ball of wool and a lot more is coming...

**Comment by jMarcS** [comment]
Too true. Here's another strand.

Maybe you caught that already: `CreateNewVar` does not really create a new `Var`. It creates a new entry in `VarNames` and `VarIndex`, but it doesn't know anything about the new `VarIndex` value. Something else has to create the `Var` and makes  sure that the new `VarIndex` item indexes it.

That's why `doUserFunction()` doesn't call `CreateNewVar()`. It's not just saving CPU cycles (for example, it does not update `VarNumber`). It's either creating the `Var`, or, for arrays and pass-by-reference scalars, copying in the index of the caller's `Var`.

**Comment by mougino** [comment]
If I read you correctly, `doUserFunction()` does not call `createNewVar()` (just checked it: yes, it seems true). So the fact that `VarSearchStart` is newly referenced in `createNewVar()` *and* that `VarSearchStart` is not updated until the end of `doUserFunction()` should have no impact at all...

I think I just need to replicate the "sorting" scheme (of `VarNames` and `VarIndex`) I implemented in `createNewVar()` into `doUserFunction()` and we should be good. How do you think?

**Comment by mougino** [comment]
Here is the fix for `doUserFunction()` :
```Java
			// Now that all new variables have been created in main name space,
			// start the function name space with the function parameter names.
			sVarNames = VarNames.size();
			for (FunctionParameter parm : parms) {
                int index = Collections.binarySearch(
                        VarNames.subList(sVarNames, VarNames.size()),
                        parm.name());                                   // index returned by binarySearch() is the place where to insert var
                index = sVarNames - index - 1;                          // so that VarNames.subList() is sorted alphabetically
				if (!parm.isArray() && !parm.isGlobal()) {
					VarIndex.add(index, Vars.size());					// new scalar
					Vars.add(parm.var().clone());
				} else {
					VarIndex.add(index, parm.varIndex());				// array or global scalar
				}
				VarNames.add(index, parm.name());
			}
```

This is not tested but I think it should fix both our problems : my call of `Bundle.Put` inside the function, and your implicit LET inside the function...
I'm away from my dev laptop. If you have time to compile the change tell me the result.

**Comment by mougino** [comment]
I just did a check : `CreateNewVar()` and `doUserFunction()` are the only 2 places in the whole `Run.java` where there are occurences of either `VarNames.add()` and `VarIndex.add()`.
With my changes (use the variant `add(index, value)` instead of just `add(value)`) the two lists should be sorted at all time thus making `binarySearch()` possible.

**Comment by mougino** [comment]
I just tried it... I still have the same crash due to an index out of bound in `Bundle.Put` :sob: 
I don't know where that comes from, I was pretty sure to have nailed it down... back to investigations

**Comment by jMarcS** [comment]
Hello, Nicolas. I have not looked at this since last weekend, I will have to start over. I do remember that there were changes needed in `doUserFunction()`, above the changes you show here. I did not work out those changes. The code for arrays calls `getArrayVarForRead()`, which indirectly calls the search function, and the code for `isGlobal` calls `getVarAndType()`.

I suppose that for the last few days you have been thinking about more important issues. I am sorry for you and your people, and for our world. Be well, mon ami.

**Comment by olimaticer** [comment]
I agree with Marc.

Gregor

**Comment by mougino** [comment]
Biggest impact of my changes is that `VarSearchStart` is newly instantiated in `CreateNewVar()` so let's focus on where `CreateNewVar()` can be called.

In  `doUserFunction()`, both the code for arrays and the code for `isGlobal` call `getVarAndType()`.
`getVarAndType()` calls `parseVar()` and `searchVar()`. Neither `parseVar()` nor `searchVar()` call `CreateNewVar()` so the problem is not in these functions.

`CreateNewVar()` is called at 3 places: in `createNewScalar()`, `BuildBasicArray()`, and `executeFN_DEF()`. The 2 last places do not concern us (not called from `doUserFunction()`).
**But:** `createNewScalar()` is called from `getVarValue()` (and only there), and `getVarValue()` **is called from** `doUserFunction()` (in its code for `isGlobal`) (and also from `getVar()`, `getNVar()`, `getSVar()`, and `ESE()`, but that's not interesting).

So the cause for the index out of bound must be in `getVarValue()`/`createNewScalar()`. I have the feeling it has to do with an inconsistent `VarSearchStart` so I'll try to find exactly why...

**Comment by mougino** [comment]
There is something I really don't understand... There are 2 separate parts in `doUserFunction()`: the first one consists in building the `parms` ArrayList, the second one scans the list and creates the variables in the function name space.

Only in the first part can we occasionally create a new var (if it is passed as parameter but never initialized before) but in this case we are still in the main name space! we do **not** need to have an updated `VarSearchStart`, on the contrary we need it to stay the one of the previous name space :worried: 

I'm really losing my latin here... Or I think we're on a wrong trail with this `VarSearchStart` thing...

**Comment by jMarcS** [comment]
Yes, that was a bugfix.

* If the function call arguments are variables that have not previously been used, then new variables must be created in the caller's variable space.
* Then, after all variables are created in the caller's space, `VarSearchStart` is moved to the end of that space.
* Then, after moving `VarSearchStart`, the arguments create new variables in the called function variable space.

The bug happened if `VarSearchStart` was moved before the first step: the previously unused variables were created in the called function space, then created again, with a different name, in the third step.

    fn.def fn(p, q) : ... : fn.end
    a = 1
    x = fn(a, b)

Variable `a` is not a problem. It is created in the caller's space. A new variable `p` is created in the called function space which is a copy of the caller's variable `a`.

Variable `b` is created new. After the bugfix, it is created in the caller's space, just like `a`, but before the bugfix it was created in the called function space. The caller has no variable `b` and the called function has a variable `b` that it should not have. Then a new variable `q` is created in the called function space that is a copy of variable `b`, where ever that is. Now what happens if the user creates a local variable called `b`? In the buggy case, there is no new variable; the function just uses the existing `b`, which is a copy of `q`. If the user changes either `b` or `q`, they both change.

`VarSearchStart` tells the interpreter which namespace it is in. The name space is always from `VarSearchStart` to the end of the variable lists.

**Comment by jMarcS** [comment]
I did not explain well at all. Let me try again.

Processing function arguments always creates new variables in the called function name space. Sometimes it also creates variables in the caller's name space -- but we (Paul and I) did not know this.

The bug: we were trying to create variables in both name spaces at the same time! This is impossible, of course.

The fix: I added the `FunctionParameter` class and split `doUserFunction()` into two parts.

First part:
As each function arguments is processed, we put everything needed to create a new local variable into a `FunctionParameter` object and put it on a list (called `parms`). This processing may create variables in the caller's name space. It does NOT create variables in the called function's name space.

After all arguments are processed, we move `VarSearchStart`. Now we can't create any more variables in caller space, only in called function space.

Second part:
We read the list of `FunctionParameter` objects that we just built. For each object, we create a new local variable in the called function space. For pass-by-value arguments, the new local variable gets the value of the argument. For pass-by-reference arguments, the new local variable gets a pointer to a variable in the caller's name space. 

The loop in the second part is much like `createNewVar()`, optimized to work with `FunctionParameter` objects and expanded to handle pass-by-reference arguments.

Anything you do to `createNewScalar` you have to do that loop in `doUserFunction()`. We've talked about that.

What we have not talked about is that expansion to handle pass-by-reference arguments.

I have taken too long; I am out of time. I will come back to this later, unless you tell me you don't need me to do that. Here is the short version:

When a variable is passed by value, the value of the outer variable is copied to a new inner variable. When a variable is passed by reference, the symbol table reference of the outer variable is copied to the new inner variable -- you don't need to look up the value, you only need the symbol table index. In other words, for pass-by-reference you need `searchVar()` but you don't need `getVarValue()`.

`getVarValue` can call `createNewScalar`. In v01.88 and before, `createNewScalar` did not need to know how to find a variable in the symbol table. It ALWAYS put a new variable at the end. With your change, that is no longer true.

The other thing `getVarValue` can call is `GetArrayValue()`, which has to evaluate array index expresssions! I have no time now to figure out how that is affected by your change. `GetArrayValue` always needs to do variable lookups, which may create new variables!

**Comment by mougino** [comment]
In your last 2 comments you talk about `createNewScalar` I think you rather mean `createNewVar` (the latter is called by the former). Anyway thanks for your explanations I had the same conclusions and my understanding was correct. The fixes I proposed should be compatible with the scheme and work, but there is an index problem anyway and I don't know where...

**Comment by mougino** [comment]
This is my test program, it does not crash but shows the wrong index problem:
```VB
fn.def test()
  let w=1
  let x=1
  let y=1
  fn.rtn x
fn.end

let y=2
print test()
print "(should be 1.0)"
```
It should print 1.0 but it prints 2.0, proving there's an issue in the variable index.

**Comment by humpity** [comment]
mougino,

Just looking at your code I see that you have done a sublist search
and corrected the returned 'index' from sublist back to mainlist offset;


createNewVar(

	index = VarSearchStart - index - 1;
------------
doUserFunction(

	index = sVarNames - index - 1;
------------



But, you have not done so in searchVar..


searchVar(

	else 
	{                                      // found it
	  j = VarSearchStart + j;	// ** don't you need this too ?? **

          if (VarIsArray)
------------

**Comment by mougino** [comment]
That's it Humpty :open_mouth: you're a genius!
So simple yet so hard to find, this solves all index issues. Binary search now works flawlessly :smile: 
Thanks!!

**Comment by mougino** [comment]
Conclusion: 3 modifications have to be passed in `Run.java`:
```Java
    private boolean searchVar(String name) {        // search for a variable by name
        int j = Collections.binarySearch(
            VarNames.subList(VarSearchStart, VarNames.size()),  // VarSearchStart is usually zero but will change when executing User Function
            name);
        if (j < 0) {                                // not in list of variable names
          VarIsNew = true;
          return false;
        } else {                                      // found it
          j += VarSearchStart;              // make the index absolute (thanks Humpty!)
          if (VarIsArray) {
            ArrayDescriptor array = ArrayTable.get(VarIndex.get(j));
            if (!array.valid()) {           // array invalidated through a different variable
              VarNames.set(j, " ");     // clear this variable so a new one with the same name can be created
              VarIsNew = true;
              return false;
            }
          }
          VarIsNew = false;
          VarNumber = j;
          theValueIndex = VarIndex.get(j);  // get the value index from the var table
          return true;
        }
    }
```
```Java
    private int createNewVar(String name, int val) {// make a new var table entry; val is an index into one of the lists
        int index = Collections.binarySearch(
                VarNames.subList(VarSearchStart, VarNames.size()),  // VarSearchStart is usually zero but will change when executing User Function
                name);                        // index returned by binarySearch() is the place where to insert var
        index = VarSearchStart - index - 1;   // so that VarNames.subList() is sorted alphabetically
        VarNames.add(index, name);                      // create entry in list of variable names
        VarIndex.add(index, val);                           // create corresponding index list entry
        VarNumber = index;
        return index;
    }
```
```Java
private boolean doUserFunction() {
// [...]
            // Now that all new variables have been created in main name space,
            // start the function name space with the function parameter names.
            sVarNames = VarNames.size();
            for (FunctionParameter parm : parms) {
                int index = Collections.binarySearch(
                        VarNames.subList(sVarNames, VarNames.size()),
                        parm.name());                                   // index returned by binarySearch() is the place where to insert var
                index = sVarNames - index - 1;                          // so that VarNames.subList() is sorted alphabetically
                if (!parm.isArray() && !parm.isGlobal()) {
                    VarIndex.add(index, Vars.size());                   // new scalar
                    Vars.add(parm.var().clone());
                } else {
                    VarIndex.add(index, parm.varIndex());               // array or global scalar
                }
                VarNames.add(index, parm.name());
            }
// [...]
```

**Comment by mougino** [comment]
And here is my BASIC! test program:
```VB
Text.Open w, fid, "../source/_VarLookupBenchmark.bas"

Text.WriteLn fid, "c0=Clock()"

For i=97 to 122
  n++ : Text.WriteLn fid, "LET ";Chr$(i);"=";Floor(65535*Rnd()+1);"  % var #";n
Next

For i=97 to 122
  Print n;"/676" % show progress
  For j=97 to 122
    n++ : Text.WriteLn fid, "LET ";Chr$(i);Chr$(j);"=";Floor(65535*Rnd()+1);"  % var #";n
  Next
Next

Text.WriteLn fid, "c=Clock()-c0"
Text.WriteLn fid, "PRINT ";Chr$(34);n;" variables registered in ";Chr$(34);";c;";Chr$(34);" ms.";Chr$(34)
Text.WriteLn fid, "PRINT ";Chr$(34);"Now accessing 500 variables randomly...";Chr$(34)

Text.WriteLn fid, "Dim timing[10]"
Text.WriteLn fid, "For i=1 To 10"
Text.WriteLn fid, "c0=Clock()"

For i=1 to 500
  If Mod(i,100)=0 Then Print i;"/500"
  Text.WriteLn fid, "dummy=";Chr$(Floor(25*Rnd()+97));Chr$(Floor(25*Rnd()+97))
Next

Text.WriteLn fid, "timing[i]=Clock()-c0"
Text.WriteLn fid, "PRINT ";Chr$(34);"Round #";Chr$(34);";i;";Chr$(34);" done in ";Chr$(34);";timing[i];";Chr$(34);" ms.";Chr$(34)

Text.WriteLn fid, "Next"
Text.WriteLn fid, "Array.Average t, timing[]"
Text.WriteLn fid, "PRINT ";Chr$(34);"Average: ";Chr$(34);";t;";Chr$(34);" ms.";Chr$(34)

Text.Close fid
Print "Done."
```
It will create a "_VarLookupBenchmark.bas" that you can run both with official BASIC! and with binary search BASIC! with my 3 modifications above.

The more entry-range the phone, the more spectacular the speed gain.
My previous results were on a very low-cost phone, here is a table with also results on my high-end phone:

Phone|Prog|Time BASIC! V01.88|Time BinSearch BASIC!|Speed gain
-----|----|------------------------|------------------------|-----------
High-range phone|_VarLookupBenchmark.bas|60|17|x3,5
High-range phone|f22_benchmark.bas|0,79|0,79|x1
Low-cost phone|_VarLookupBenchmark.bas|250|52|x4,8
Low-cost phone|f22_benchmark.bas|2,12|0,2|x10,6

**Comment by humpity** [comment]
Just a suggestion, but I think there should be some sort sanity check in douserfunction

    if ( index<0 )
      {  index = sVarNames - index - 1; }
    else
      {  <error> }

I know it doesn't crash (01.88.02) if the user types in something silly like
fn.def myfunc1 (a,a)
because it just adds the second 'a' to the end of the list.
But a b.search index might crash.

**Comment by mougino** [comment]
Just tested it with binary search: `fn.def myf(a,a)`, `fn.def myf(b,a,a)`: it doesn't crash neither.
I am looking why. When we have an explanation and we are sure this part of `doUserFunction()` is exception-free (=crash-free), we can commit like I proposed.

**Comment by mougino** [comment]
My previous comment was slightly incorrect: `doUserFunction()` treats the call to a function, not the definition of a function (which is done by `executeFN_DEF()`).

Even so, with a `fn.def myf(a,b,c)` and a `call myf(b,a,a)` binsearch BASIC! doesn't crash. Still investigating why...

**Comment by mougino** [comment]
I put some traces in `createNewVar()` and `doUserFunction()` and here is the result for the following test program:
```VB
fn.def test(a,b,a)
  let z=5
  let x=1
  let y=17
  fn.rtn x
fn.end

let y=2
print test(b,a,a)
print "(should be 1.0)"
```

First, variable `test(` is created by `createNewVar()` at index 0/0 (search-start=0)
Then, variable `y` is created by `createNewVar()` at index 1/1 (search-start=0) (this is the `let y=2`)
Then, variable `b` is created by `createNewVar()` at index 0/2 (search-start=0) (this is the `print test(b,a,a)`)
Finally, variable `a` is created (**once**) by `createNewVar()` at index 0/3 (search-start=0) (normal, the second call is a reference to the already existing var)

Then we enter the function:
Function variable `a` is created by `doUserFunction()` at index 4/4 (search-start=4)
Then, function variable `b` is created by `doUserFunction()` at index 5/5 (search-start=4)
**Then, another function variable `a` is created by `doUserFunction()` at index 3/6 (search-start=4)** :warning: 
Then, function variable `z` is created by `createNewVar()` at index 7/7 (search-start=4)
Then, function variable `x` is created by `createNewVar()` at index 7/8 (search-start=4)
Then, function variable `y` is created by `createNewVar()` at index 8/9 (search-start=4)

So you are right Humpty, the line in bold shows there is a problem of creating a variable outside of the function name space.

@humpity @jMarcS what kind of error should we throw if user does a `fn.def myf(a,b,a)`? Something like `Error: duplicate parameter name`?
Or should we handle it smartly (skip the second parameter `a` which already exists) and not throw any error??

**Comment by humpity** [comment]
yes, the 2nd 'a' causes a 'hit' which does not return a negative index. Such hits will cause the index
to go the other way back up to corrupt the caller's namespace;
 index = sVarNames - index - 1;
 = 4-(0)-1 = 3 (when search-start=4)
.( don't know what was at index3, possibly a function name?, anyway, it got stepped upon ! )
I think yes, "duplicate parameter name" and it should throw an error. It's a silly user error which he could have
avoided.

---------

Also, you don't need the 'else' in searchVar;
e.g

    private boolean searchVar(String name)         // search for a variable by name
    {
        VarIsNew = true;			// assume new variable

        int j = Collections.binarySearch(
            VarNames.subList(VarSearchStart, VarNames.size()),  // VarSearchStart is usually zero but will change when executing User Function
            name);

        if (j < 0) return false;            // not in list of variable names, must be new
          
                                            // else found it
        j += VarSearchStart;                // make the index absolute
        if (VarIsArray) 
        {
          ArrayDescriptor array = ArrayTable.get(VarIndex.get(j));
          if (!array.valid()) 
          {           // array invalidated through a different variable
            VarNames.set(j, " ");     // clear this variable so a new one with the same name can be created
            return false;		// must be new
          }
        }
        VarIsNew = false;
        VarNumber = j;
        theValueIndex = VarIndex.get(j);  // get the value index from the var table
        return true;
    }

**Comment by mougino** [comment]
Great!
Then I propose the following [trick](http://stackoverflow.com/questions/562894/java-detect-duplicates-in-arraylist) to detect a duplicate parameter... and we should be done :wink:

**Comment by mougino** [comment]
@jMarcS I am ready to commit, how do you stand on these modifications?
Do you want I send you the modified `Run.java` by mail and you review it during the week-end maybe? (and I leave you control over possible modifications and final commit?)

**Comment by mougino** [comment]
I opened issue #200 for the duplicate parameter problem, where I propose a fix (tested, works fine).

**Comment by humpity** [comment]
Isn't it far simpler to catch the error (+ve index) in douserfunction ? (also serves as a sanity check).
Fn.Def is the proper place, yes, but it's such a rare error..

**Comment by mougino** [comment]
I tried it before :wink: but `doUserFunction()` treats the call to the function, so if you throw a run time error in `doUserFunction()` it will report an error in `print myFunc(1,2,3)` (where there is none)

**Comment by humpity** [comment]
That's true!
Maybe a b.search in Fn.Def too? Would that be better than an extra hashset ?

There's also the issue that if Fn.IMPORT is implemented, then some sort of test has to be done also in douserfunction.

**Comment by humpity** [comment]
> Maybe a b.search in Fn.Def too?
Forget that, the ordering has to be preserved.

**Comment by jMarcS** [comment]
humpty, you caught a potentially really bad bug. Thank you.

With the old variable creation scheme, a duplicate parameter name was a subtle bug, but not very serious. I haven't tried to prove it, but I think the effect is that the duplicate parameter was just ignored. Two variables are created with the same name, but the linear search could find only the first one.

With the new scheme, you wind up with an index before `VarSearchStart`, potentially less than 0. The caller's space is corrupted, or the interpreter crashes. Not good.

This happened because `doUserFunction()` creates variables without first checking to see if they exist. I sincerely hope that this is the ONLY place in the code that creates variables without checking!

Nicolas has a nice fix for this; both cheap and easy. But I feel humpty's unease. If we're going to program defensively, then:
- ANYTHING that creates a variable should go through a common routine, maybe `createNewVar()`.
- That common function should protect itself: it should check if the variable to be created already exists.

**Comment by jMarcS** [comment]
I was concerned about the pass-by-reference cases. I have convinced myself that Nicolas's code handles them correctly. I think I was confused about `VarNames`, `VarIndex`, and `Vars`. The new scheme inserts items into the first two, but we still add items only at the end of `Vars`.

I also ran some parameter-passing tests and they all, um... passed (pun not intended, but enjoyed nonetheless!).

**Comment by humpity** [comment]
My concern was if the error merited an extra overhead in Fn.Def.
createNewVar is a great choice. It will have the new b.search so it will know right away if the name already exists
without creating a hashset.
So (for this issue) it just remains what kind of error msg will be given?

**Comment by jMarcS** [comment]
>  if the error merited an extra overhead in Fn.Def.
Absolutely. `Fn.Def` runs only once per function, so a little (more) overhead is okay.

I suppose we could make it a little better. We don't have to create a HashSet unless there are at least two parameters. But that's just the uber-geek in me. (At least two parameters of the same type! But that way lies madness...)

> createNewVar is a great choice. It will have the new b.search so it will
> know right away if the name already exists without creating a hashset.

I think you're right. Overhead in `createNewVar` is more expensive: it runs once for every variable name, including array names and function names. Not nearly as bad as overhead in `varSearch()`, of course. It's probably small, and I think it's worth it. Still, I think we'll need a performance measurement to be sure.

Somehow we have to combine `createNewVar` and the changes in `doUserFunction()` so that the binary search call appears only once. That's where the test for a positive index value should go.

>  what kind of error msg will be given?

If the rest of the code is done right (and I think it is, with Nicolas's fix), you will never get this error. It could be of the "Notify developer!" type that we already have for some other errors that indicate the interpreter code is broken. If not that, then something like "Attempt to create variable " \<variable-name\> " that already exists". Opinions?

One other consideration: I'm not certain we can call `RunTimeError()` from `createNewVar()`. If not, reporting the error gets ugly, because its caller has to check.

We really really need a calling tree for the components of `GetVar()`! I used to use `egypt` to draw call graphs from GCC symbol (RTL) files. It's a Perl script that walks the symbol table looking for connections and sends the result to `graphviz`. I wonder if there's anything like that for Java?

Perhaps I'm over-complicating this issue -- I've always felt one of the most important things an engineer can do is worry about things that could go wrong.  But I do think it will take a little longer to work through this stuff and I don't want to hold up the binary search change any more.

I think it's time to commit Nicolas's code. I have some pretty-printing suggestions and a comment change or two, but nothing algorithmic.

**Comment by jMarcS** [comment]
My wife is on the phone with a friend, so I had a little time to kill.

I added an index check to `createNewVar()`. I had to add a duplicate check in `doUserFunction()`.

    if (index >= 0) { throw new InvalidParameterException("createNewVar: variable " + name + " already exists"); }

That won't give a Runtime error to the BASIC! user. It throws a Java exception. I think this is a good way to handle internal errors.

Well, apparently I have thought that before, because there are a bunch of other places I throw `InvalidParameterException` for internal errors. (Also some places where I catch that Exception, maybe I should stop catching them?)

Since that has somehow become the de facto way to catch internal errors, I added a clause to the `UncaughtExceptionHandler()` for `InvalidParameterException`. Should have done that a long time ago.

To test it, I took out Nicolas's `HashSet` test and ran a BASIC! program with a function that has duplicate parameters. The program (not BASIC!) crashes, and I get a stack trace of the duplicate variable creation attempt. Down at the bottom,  it does the standard Runtime Error thing and prints the offending line of the BASIC! program. Beauty!

What does the index check cost? I don' t know. I tried to run Nicolas's performance test  on my old gingerbread phone. I can't get results consistent enough to say if there is a measurable difference in performance. There's something else going on, maybe some Android activity, that skews the results of any run by tens of milliseconds. The fastest batch I saw says that adding the index test makes the code run faster. That's impossible, so either I have to find some way to clean up the interference, or I have to declare the results a draw: adding the index check is essentially free.

**Comment by jMarcS** [comment]
I am not able to achieve Nicolas's performance results.

I get a nice improvement in the variable-heavy test program _VarLookupBenchmark.bas. On my Moto G phone it's 4x or 5x improvement.

But I see only a tiny bump in speed of the f22_benchmark. It's about 2% (not 2x) on an old XOOM tablet running Jellybean (2.30 -> 2.25). On my Moto G it may be about the same, but 2% of a 0.80 second-run is noise.

Any ideas?

**Comment by humpity** [comment]
Perhaps because f22 uses just a few variables?

Wikipedia:  https://en.wikipedia.org/wiki/Binary_search_algorithm#Average_performance
"In theory binary search is usually faster than linear search, but in practice that may not hold true. For small arrays (say about 64 items or less), linear search may have better performance. "

Also: https://schani.wordpress.com/2010/04/30/linear-vs-binary-search/

**Comment by jMarcS** [comment]
Yes, I'm sure you're right, but up in that nifty little table Nicolas posted last week (and how do you post a table to GitHub anyway??), it says 10x for a low-end phone, and no improvement for a high-end phone. I'm seeing no improvement at either end. I don't have a high-end phone, but my Moto G is fairly recent, and reports f22_benchmark numbers in the 0.8 second range).

Looking more closely, I wonder if that's an error. I believe 0.79 for the high-end, but I don't think 0.2 for a low-end phone is right. My numbers for _VarLookup match the table well enough (2-3x for the slow XOOM, 4-5x for the fast phone).

This is off-topic, but I thought I'd mention something else I see. _VarLookup does 10 "rounds" and the variation between them is huge. The tablet just goes away for a while, anywhere from 200 to 500 ms. I think it's about the same for either version of BASIC!, so it tends to weaken the performance improvement measurement.

I'm sure an analysis of what's it's doing would be interesting. I'm pretty sure BASIC! does way too much garbage collection, even when it's just sitting in an idle loop. I've noticed it before, and I don't know what's causing it. I know there are changes in Lollipop to smooth out the GC, but it's still doing work it shouldn't need to do. But that will be another project for another time.

**Comment by humpity** [comment]
HTC Sensation |||
-------|----------|------|---------
_VarLookup | 235.1 | 61.4
 F22 | 2.5166 | 2.4579


( will try old phone tonight )
( https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables )

**Comment by humpity** [comment]
Summary

HTC Sensation

ICS 4.0.3|01.88.02 | b.search|diff
----------------|--------|--------|---
_VarLookup|235.1|61.4|3.82x
f22|2.5166|2.4579|1.02x

LG P500

Froyo 2.2.1|01.88.02|b.search|diff
----------------|--------|--------|---
_VarLookup|1201.2|186.0|6.45x
f22|8.943|7.575|1.18x

**Comment by humpity** [comment]
Can't make head or tail of this, but Collections binarySearch shouldn't really work for Froyo. After a bit of googling, the Collections binarySearch only came in since Gingerbread (API 9 - 2.3). The Array binarySearch was always there (API 1).

**Comment by jMarcS** [comment]
Thanks for posting your test results. (And for telling me how to do tables!) The results look pretty consistent.

There must be some misinformation floating around out there. I went to the authoritative source, `developer.android.com/reference`, and it looks like we're okay. This is a Good Thing; I didn't test on a Froyo emulator before committing.

There are nine flavors of `BinarySearch` in the `Arrays` class. Each of those has two variants. The full-list variants have been around since API 1. The sublist variants showed up in API 9. From the [Arrays page](http://developer.android.com/reference/java/util/Arrays.html):

    public static int binarySearch (double[] array, double value)                                      Added in API level 1
    public static int binarySearch (double[] array, int startIndex, int endIndex, double value)        Added in API level 9

In the `Collections` class, there are only two signatures, and they've both been there since the beginning. From the [Collections page](http://developer.android.com/reference/java/util/Collections.html):

    public static int binarySearch (List<? extends Comparable<? super T>> list, T object)                   Added in API level 1
    public static int binarySearch (List<? extends T> list, T object, Comparator<? super T> comparator)     Added in API level 1

**Comment by humpity** [comment]
Yes, I got Arrays and Collections mixed up.
Putting aside the b.search, another theory I have could be the effect of 'getter/setter' inline optimizations in the
Dalvik engine after Froyo.
http://stackoverflow.com/questions/4912695/what-optimizations-can-i-expect-from-dalvik-and-the-android-toolchain/4930538#4930538
This might make a linear search faster after Froyo, although I don't know if you can optimize VarNames.get(j).
Anyway, I think I'm getting off topic now.

**Comment by jMarcS** [comment]
That link made interesting reading. Thank you.

There are two linear searches. No, *were* two linear searches -- Nicolas's change makes variable name searches binary. That's because a variable name can be extracted from the command line as a discrete token, then the token can be used in a name table lookup. I had tried a hash table, but I needed a new table for each level of function call. Nicolas found that BinarySearch can be applied to a sublist, and that has made all the difference.

The command keyword search is still linear. That's because the keyword can't be isolated as a token. Instead, the beginning of each line is compared to every entry in the command keyword table (stops when a match is found). (I know, I've said this before, but it's a necessary preface to what comes next).

That compare looks like this in pseudocode:

    for each keyword in the table
      if command line starts with the keyword
        table hit

The Java method is `String.startsWith(...)`. And it is *fast*. Blindingly fast. Faster than it has any right to be. So fast that everything I've tried to make it faster makes it slower!

So I quit trying. Instead, I changed the data structure that holds the commands so the keyword search happens only once per line. Originally, each line was just a Java `String`, and it got fully parsed every time the interpreter saw it.

Now I have a table of command functors, one for each command keyword. The first time the interpreter sees a line, it identifies the keyword and stores a reference to the functor. Each command line is an object that holds the original `String`, a functor reference, and the length of the keyword. When the interpreter sees the command again, it doesn't parse anything. It grabs the functor reference, skips ahead as far as the length field says it should, and starts the command.

I don't remember the speed-ups that got. Also, I did it in two stages. The first stage did non-compound commands (no dot). The second stage (much harder!) handled compound commands (commands in groups, like `GPS.latitude` or `GR.bitmap.load`. I've never tried to measure the combined effect.

If I remember right, I was a little disappointed in the effect. Expression parsing is a bigger concern, and variable name lookup is the slowest part of that.

The first stage went into v01.85. The second went into v01.88. The BinarySearch change will be in v01.89. There are hundreds of other changes between v01.85 and now, but it would still be interesting to compare benchmarks.

And now that the "search" part of variable-name handling is faster, I'd like to do some profiling to see where the next hang-up is. Probably the string extraction. If that's the case, the next speed-up will be tough -- pre-parsing the line, saving the tokens, maybe the expression tree? Putting even more "compiler" into the "interpreter".

I have to wonder if the time would be better-spent trying to find the massive memory leak. Or getting the interpreter out into a Service where it belongs.

But now I'm WAAAAAY off-topic!

**Comment by jMarcS** [comment]
**Okay, back on-topic.** Gentlebeings, we have a bug. `f25_dir.bas` fails. You're gonna love this (note the sarcasm, please).

`UNDIM` (same as `ARRAY.DELETE`) works by replacing the variable name with a space character. (`searchVar` does the same thing, to propagate the effect of `UNDIM` on an array passed into a function.) The intent was to write an invalid variable name so that future linear searches would skip it. This "removes" a variable without changing the index of any other variable. Then `DIM` (or something else) can make a new variable with the same name and a new index, and the linear search would find that. 

With `BinarySearch`, writing over a valid variable name with a space is evil: the list is no longer in alphabetical order. Furthermore, multiple `UNDIM`s would create multiple variables with the same name (" "), which is not allowed with `BinarySearch`.

On the other hand, with `BinarySearch` the variable indices can change every time we add a new variable. Nobody cares, because the index values are not directly accessible. Every time the variable is looked-up, the search gets the current index.

I thought that might mean deleting variables is okay now. That could be a nice clean fix. But interrupt code can `UNDIM` an array in the wrong namespace, invalidating all of the `VarSearchStart` values saved in the call stack.

I thought we could "rename" the variable by cloning it with a new name, inserting it with `BinarySearch`, and deleting the old one. We'd have to create a naming scheme that would avoid duplicate invalid variable names. Besides, this is an opportunity to right an old wrong: adding unused variables to the list just slows everything down.

We could fix this by adding bookkeeping to the array descriptor. `UNDIM` would not change the variable, but any variable look-up that found an `UNDIM`ed array would report the variable does not exist. An attempt to create a new array with the same name would not change the variable either, except to point it at a newly-created array descriptor. It's a little complicated, but I haven't thought of a reason it would not work. And it removes the extra step in `searchVar` to propagate `UNDIM` out of function calls. And it might be movement in the direction of separating arrays from scalars -- not sure about that one, it's hard to think about.

Any other ideas? I'd like clean and simple, but I'm moving in the opposite direction.

**Comment by humpity** [comment]
I can sympathise, it's one of those programming moments that makes your stomach
turn inside-out.

If I read you right these are the following scenarios with UNDIMed vars (uvars);
```
A) Pretend nothing happened.
   Clearly does not work as replacing uvars with a 'space' screws up b.search.
   Not an option.

B) Replace uvars with a naming scheme that moves uvars to either the top
   or bottom of the list.

   (without invalid duplicates)
   Inventing a scheme e.g " myname123" can prove messy.

   An easier way, would be to just insert a space " myname"
   If the user or system tries to DIM "myname" again, then 
   " myname" can first be searched for and " myname" replaced with "myname"
   to be re-used. In a way it is similar to the BookKeeping method.

   (with invalid duplicates)
   I'm not yet convinced that Inserting or replacing the name with a space
   AND moving it to the top of the list would neccessarily screw up a
   b.search.
   This method though does not 'right an old wrong',
   you still end up with duplicates, albeit 'safe' duplicates.

   For both cases (invalid or valid) I'm not sure how it affects undim
   propagation in functions.

C) BookKeep with array-descriptor, i.e marking uvars, making uvars re-usable.
   Simplest option ?

D) Really delete variables.
   
   This option only works if VarNames, VarIndex and the various VarSearchStart's
   are all synchronised properly.

   This means that after deletion, UNDIM must somehow scan the call stack entries 
   and correct all the VarSearchStart's depending if a particular VarSearchStart
   was before or after the deletion (can this be done?).
```

**Comment by mougino** [comment]
I'm on smartphone, so I may have read Marc's post in diagonal, however: how about removing the arraylist element in VarNames and VarIndex (only) altogether, instead of filling with a space? 
Wouldn't that work?

**Comment by mougino** [comment]
Ah right, I read a little more in detail... 
How about throwing an error "Cannot undim in an interrupt" then?
It's a new limitation but would solve all of our problems :white_check_mark:

**Comment by jMarcS** [comment]
Yes, Nicolas, that's a very tempting suggestion.

Humpty, I think your recap is mostly right on. I looked at Option C a little, and it's worse than I thought. For Option B, as you suggest, there's probably a way to deal with the duplicate variables; true duplicates would have to be excluded from the sublist being searched.

It annoys me that we have to pretzel the code to account for array names! There's got to be a better way.

But -- I've got to get the test build out so users can check out both the speed-up and the new Paint features. Maybe the right answer is to do quick-and-dirty for now and fix it right when we have more time.

Quick-and-dirty would be Option D with a mouginish twist:

- in both places that the variable name is blanked, delete the variable (both `VarNames` and `VarIndex`.
- Do not delete the array descriptor.
- In `UNDIM`, if the index of the variable being deleted is smaller than `interruptVarSearchStart`, throw an "undim in interrupt" error. (This will have to be cleaned up a little; I was sloppy with `interruptVarSearchStart`.)

This is really easy.

If we do it, it will lean us toward deletion as the permanent solution. Walking the call stack and decrementing each saved `VarSearchStart` is easy, but it feels awfully kludgy. I want to make the code less fragile, not more!

If we could have different variable lists for each function call, the call stack would save a reference to the variable list instead of the start of the sublist. Can't do that until the variable list is a single list. But maybe we could leave the kludge in place until then?

**Comment by humpity** [comment]
Does UNDIM do the propagation? OR does it occur at FN.RTN/FN.END ?
i.e After function2 UNDIMs, at which point does function1 know when to UNDIM it's own variable ? after the
exit of function2 or before ?

**Comment by jMarcS** [comment]
Neither. `UNDIM` marks the array descriptor invalid and clears the variable name. The function returns and the interpreter goes on running. If it gets an instruction that references that array, the variable lookup finds the invalid array descriptor and clears that variable name, too.

It's in `SearchVar()`, there where it is says `if (VarIsArray)`.

**Comment by humpity** [comment]
I see. The bookkeeping is sort of already there.
Then I think this would work. The ony thing is that you can't undim a 'main' variable in an interrupt, there might be
some negative feedback from users who have done this (can't be many).
(btw I don't see the kludge as a 'kludge', as it ensures call stack integrity)

An alternative quick fix is to 'always' re-insert a space at index 0 (Option B-b) if index < interruptVarSearchStart,
but that leaves the problem of unused entries).

**Comment by jMarcS** [comment]
Nicolas reported in the forum that the GW-demo doesn't work with the v01.88.03 test build.

First look: we're running into trouble with all those global variables: `theVarIndex`, `VarNumber`, etc. For example, in `getArrayValues()`:

    int avn = VarNumber; boolean avt = VarIsNumeric;
    do { evalNumericExpression(); [...] } while (isNext(','));
    VarNumber = avn; VarIsNumeric = avt;

That is, `evalNumericExpression()` is going to change `VarNumber`, so grab a copy. Then, after evaluating the expression, restore `VarNumber`.

Yeah, that was fine when a variable's index in the variable lists never changed. Now creating a new variable can move any or all of the other variables. Saving/restoring the index doesn't work any more.

For example, this crashes:

    a=1 : c=3
    array.load b[],11,12,13
    d=b[a + b]   % d=b[a] does not crash

How many places are there like this in the code, where it assumes that a variable's index never changes? I don't know, but it looks like I'm going to have to find out.

**Comment by jMarcS** [comment]
That commit I made a few minutes ago says "Tiny fix in GetArrayValue()". I moved one line, so it gets the `ArrayDescriptor` before evaluating the index values. This way it doesn't care if `VarNumber` gets restored correctly, and my little code snippet doesn't crash.

Of course, `VarNumber` and `VarIsNumeric` still are not getting restored correctly. I suppose that's a problem, but I haven't got a small test case to prove it.

The fix in `GetArrayValue()` does not fix the problem Nicolas found testing with GW Lib. Or maybe it does fix one problem, but `GW_demo` still crashes. The error in the stack trace is "Index out of bounds". It's trying to reference `Vars`, and the index it's using is the same as the size of `Vars`.

I tried taking out the `UNDIM` change, and it made no difference. The `UNDIM` change removes an item from `VarIndex`, not from `Vars`.

I could, and probably will, go on trying to find the cause of that indexing error. But in the meantime, I'm starting to prepare for a more stable fix. The first step is in that same commit: "Changed Var to Val and Vars to Vals". That's the class `Var` and the `ArrayList<Var> Vars`. The terminology is very confusing: things I've been calling vars in the code are really objects that hold the value of a scalar. The change looks very large: 544 additions, 537 deletions (counting the insignificant change in `GetArrayValue()`). It was mostly mechanical, global search and replace. It has no effect at all on the program logic, but it makes the thinking logic easier.

With that name change, I can add a new class called `Var` that will hold the name and type of the variables -- yes, I intend to replace `VarNames` and `VarIndex` with a single list of `Var` objects. `VarIsNumeric`, `VarIsArray`, `VarIsFunction`, and `VarIsNew` will all vanish, and good riddance. And each `Var` object will hold a reference to a `Val` (for scalars), an `ArrayDescriptor`, or a `FunctionDefinition` -- just the same as `VarIndex` does now, except it will hold *references*, not *indices*. That may (should?) mean we can delete scalars. And `UNDIM`ed array space.

I haven't thought through what this does to `binarySearch()`, yet. Have to add a `Comparator`, and then see if that slows down the variable search. I don't want to undo Nicolas's gains!

This is not going to look exactly like the "Grand Unified Variable" architecture I've been working toward. I just have to hope it's a little easier to implement than that would have been.

**Comment by olimaticer** [comment]
Some weeks ago Marc wrote about a database approach.
Now I'm thinking. Marc looks more and more in this direction. 

And now I know why.
Now I now why the guys from CA (Realizer-Basic) chose another approach.

To fence the spaghetti code of GW-Basic they made it simpler as VBASIC.

First all variables are global except which are in function or procedure calls or are declared 
as local in them. (In Basic! you could call „GlobalsOn“.)

Functions or procedures have to be on top of the code except commands like the RESET command.

So the Realizer approach need only one reference table for (keywords, operators,) variables, functions and procedures, but no different name space.  
OK, hit me, in this example I use name extensions.

```
FUNC MyFunc (s) REM Only single declaration allowed. 
	LOCAL b, c
	a = a + 5
	c = 1
	b = s + c + a
	p = „Test“
	RETURN b
END FUNC
s = 10.3
c = 6/4
TYPE a AS Integer REM Only single declaration allowed. 
a = INT(2.0)
p = 99
PRINT MyFunc (33.0) → 41.0
PRINT s, c, a   → 10.3   1.5   7
PRINT p → Test
```
		Reference table records step by step: (The first item is the key.)
MyFunc | Function | start | line1
MyFunc#s | ? | Free | ?		REM Swap s with MyFunc#s.
MyFunc#b | ? | Free | ?
MyFunc#c | ? | Free | ?
a | ? | Free | ?
MyFunc#c | Integer | Free | 1
MyFunc#b | Integer | Free | ?
p | String | Free | „Test“
MyFunc#RETURN | ? | Free | ?
MyFunc | Function | end | line8
s | Double | Free | 10.3
c | Double | Free | 1.5
a | Integer | Fixed | ?
a | Integer | Fixed | 2
p | Integer | Free | 99
MyFunc | Function | used |
MyFunc#s | Double | Free | 33.0
MyFunc#b | Double | Free | 41.0
p | String | Free | „Test“
MyFunc#RETURN | Double | Free | 41.0



The reference table could be
:

Key ((keyword, operator,) function name or variable name) 
| 
Type (function, keyword, integer, double, string, bundle, array of strings …)
|
Variable type fixed?
|
Adress or Value ( small values, pointers,  branch destinations )

This table is a case for binary search or may be faster Java search methods.

With your own binary serarch you get the point on which you can split the table, if you want to insert an new keyword. So the table is in the right order without starting a new sort process.
(See my Binary Search enhancement at the rfo-basic-forum also.
http://rfobasic.freeforums.org/post25410.html#p25410 )

Some remarks:
Realizer uses bundles like a.Name, a.PostalCode, a.City, … .
The most variables are able to change there types at runtime!

As a beginner you hate it, then you love it and later you use the LOCAL command consequently.

If the source code has reached a larger size and for every include file, Realizer generated an object code at first runtime and save it on the disk. I suggest *.bao as a file name extension. Ballast like unused functions or remarks could be thrown away. The reference table could stored in in this file, too.

I do not claim, this approach is better, but he is different and well-considered.

On the other hand, some months ago I was looking in the Java subs. So I seam to remember, all the variables are stored in Java as text. I´m right?

I hope this post gives some inspirations.

Gregor

„Make things as simple as possible, but not simpler.” (Albert Einstein)

**Comment by jMarcS** [comment]
I can't begin to describe what I've been doing for the last nine days. Some day I will have to do that, but not now. While testing those changes, I found a very small but very important error in the original "binary search" commit:

    // Now that all new variables have been created in main name space,
    // start the function name space with the function parameter names.
    sVarNames = VarNames.size();
    for (FunctionParameter parm : parms) {
        // Get insertion point (-index - 1) so VarNames.subList will still be in alphabetical order.
        int index = newVarIndex(parm.name(), VarNames, sVarNames);

That last line is the common function that uses `binarySearch` to find the insertion point of a new variable in the (sorted) variable list. The error is painfully obvious in hindsight: we're capturing the end of the caller's variable space in `sVarNames`, but we don't change `VarSearchStart` until a few lines later. Oops!

    sVarNames = VarNames.size();
    VarSearchStart = sVarNames;
    for (FunctionParameter parm : parms) {
        int index = newVarIndex(parm.name(), VarNames, sVarNames);

This fails in any program that uses a function parameter name that matches the name of a variable that already exists in the caller's name space. It's amazing that so many tests worked!

I'm committing a change that fixes this little oversight. It also re-enables hardware graphics acceleration in HTML mode, as Nicolas requested over in Issue #196.

Unfortunately, this commit still doesn't make the original binary search speed-up code work. `GW_demo` still gets an `IndexOutOfBoundsException` just off the end of the `Vals` list.

**Comment by jMarcS** [comment]
I have a candidate for check-in. It runs GW_demo.bas. It runs all of the Sample_Programs (except I didn't run f32 or f35). That's not nearly enough.

My changes touched almost every operation in the code. They affect every command. Ideally, every command must be tested, including tests for syntax errors. Needless to say, that isn't going to happen. It's next to impossible.

If I had any sense at all, I would have tabled Nicolas's binary search code a week ago. But I didn't, so now we have to make a choice.

1. Remove the binary search code, release v01.89 without it. Put it in when we have time to fix it.
2. Commit my new code. Build v01.88.04. Deal with the errors as they turn up.

I fixed the errors that were easy to find. Now it gets harder. If we could get some forum users to run their favorite programs, we can debug it well enough to release it.

I really want to release this code. I have put too much effort into it to drop it. But we must get v01.89 out, and there's not enough time to test the new code as it should be tested.

What do you you think?

(BTW, I have not been in the forum since 12/4. I am waaaay behind. Anything you post there I will not see for a while.)

**Comment by mougino** [comment]
Wow, thanks Marc for your efforts! I will be interested to know what the fix was eventually.
In the meantime I think the forum is a great place to have your fix tested. Many active users for many kinds of programs. 

I don't know what is your timeline for v01.89, if you really want to see it out asap I would suggest getting rid of the binary search and releasing it on the Google store with the rest of the changes. Then we will have all the time to test a v01.89.01 with only the binary search.

**Comment by jMarcS** [comment]
Yeah, well, my timeline for v01.89 was mid-October. Didn't make it!

I think the only change I have not yet made that absolutely must go into v01.89 is the Save path bug.

Let me look at how to back out the binary search while still keeping the Paint and other recent changes.

**Comment by jMarcS** [comment]
I will attempt a brief explanation of the changes. Explaining them is not hard. Keeping it brief is hard!

The values list (`Vals`) is gone. The globals `VarNumber`, `theValueIndex`, `VarIsNew`, `VarIsNumeric`, `VarIsArray`, `VarIsFunction`, `ArrayTable`, ArrayValueStart`, 'FnDef`, `WatchVarIndex` are gone. Function names are global, so I had to keep `FunctionTable` -- one step removed from having global variables!

`VarIndex` was a list of indices into other value lists (`Vals`, `ArrayTable`, `FunctionTable`). That's gone, too, replaced by `ArrayList<Var> Vars`. `Vars` is a list of variables. Each `Var` object is a complete variable: a scalar, an array, or a function definition. `Var` is an abstract class with concrete subclasses `ScalarVar`, `ArrayVar`, and `FunctionVar`.

Each `Var` object contains a `Val` object to hold its value. `Val` is an abstract class with concrete subclasses `NumVal`, `StrVal`, and `ArrayVal`.

A `ScalarVar` holds a `NumVal` or a `StrVal`. An `ArrayVal` holds both an `ArrayDef` and an `ArrayVal`. A `FunctionVar` holds a `FnDef`. A `FunctionVar` does not have a value (no `Val` subclass`).

An `ArrayDef` holds a Java array. Yes, a real Java array, although for now it is still only one dimensional. `UNDIM` discards the `ArrayVal`, releasing all of the array storage. (Supposed to -- I haven't tested that yet.)

`Var` and `Val` and their subclasses, along with `ArrayDef`, `FnDef`, `FunctionParameter`, and the `Type` enum are all in the file `Var.java`. It's right around 500 lines, although `Run.java` is only about 300 lines shorter.

The symbol table is still in two parts. The variables are in `Vars`, but I kept `VarNames`, the list of name strings, because `binarySearch` uses a fast native (C++) search function for strings. (Maybe some day we could combine `VarNames` and `Vars` into a single `HashMap<String, Var>`.)

This touches all of the `GetVar()` derivatives and variants, but they really have not changed that much. The set of methods is the same, which makes the change a lot more manageable. Instead of passing around a `String` and setting a lot of globals, the methods pass around a `Var` that knows what kind of variable it is. That makes the whole system a lot more stable.

For performance, `ParseVar()` does not create new `Var` objects. It reuses a pool of three, one of each of the `Var` subclasses. This should mean a lot less garbage collection in idle loops. We have to be careful that the `Var` returned by `ParseVar()` doesn't get used as a real variable. `SearchVar()` uses it only for the variable name. If it finds an existing variable, it returns the existing `Var` from `Vars`, not the `Var` from `ParseVar()`. Anything else that calls `ParseVar()` either clones a new `Var` or uses the information from the `Var` without passing the `Var` itself.

The old `getVarValue()` worked by setting the global `theVarIndex`. All of the 100+ places that call `getVar()` or `getNVar()` or one of the other variations used `theVarIndex` (and `VarIsNumeric` and the other globals) to find the actual `Val` to read or write. Even `GetArrayValue()` set `theVarIndex` to an index of a value, because an array element was just another scalar in the list of scalar values. All of those methods return a `boolean` to say if they succeeded; if they returned `true` the real "return value" was in `theVarIndex` (and `VarNumber` and `VarIsNumeric` and so on).

To reduce the size of the change, I kept that behavior. The methods still return a `boolean`. But now all those globals are replaced with a single `Val` reference, the global `mVal`. (Some day, that will go away, too, but that will be a pretty big change all by itself.)

Getting a `Val` reference was easy for scalar variables. It was much (much!) harder for arrays. In Paul's design, an array's data was a section of the scalar values list. (Or lists: in Paul's original design there was a list of String objects and a list of Double objects. Some time ago I combined them into one list of objects of a class that could be either.) An `ArrayTable` element described an array by keeping the index of the `base`, where the array started, and a `length`.

Now there is no `ArrayTable`, just individual `ArrayVar` instances. Each `ArrayVar` has an `ArrayDef` which holds the data in a real Java array. All of Paul's expression parsers rely on treating an array element exactly the same way as a scalar. In the new design, they have to have a `Val`.

So I created a new `Val` subclass, `ArrayVal`. Each `ArrayVar` has both an `ArrayDef` and an `ArrayVal`. The `ArrayDef` keeps the dimensions and all of the data. The `ArrayVal` is a temporary write-through cache for one element of the array.

If you have `a = b[x,y]`, `executeLet()` calls the numeric expression parser, which calls `getNVal()` to get the `ArrayVal` that caches an element of the array `b[]`. The `ArrayDef` that holds the array data uses the index list to select an element of the array, then writes both the index and the array element value into the `ArrayVal` cache. The expression parser finds that `ArrayVal` in `mVal` reads the value -- it is not reading the array directly, it is reading the cached value.

If you have `b[x,y] = a`, `executeLET()` calls `getVar()` directly to get the `ArrayVal` that caches the element of the array `b[]`. The process and the result are exactly the same: `mVal` is set to the `ArrayVal` that caches an element of the array in the `ArrayDef` attached to the `ArrayVar` of the array `b[]`. The cache holds the current value of the array element `b[x,y]` -- but nobody cares, because this code wants to write the array, not read it. `executeLET()` calls the numeric expression parser which calls `getNVal()` to get the value of the scalar `a`. It writes the value of `a` to the `ArrayVal` cache. The `ArrayVal` is a *write-through* cache. It has the index and a reference to the `ArrayDef`, so it writes the correct element of the original array to the new value.

There is no cache invalidation mechanism. Because any BASIC! array read or write is done with a full set of indices every time, the cache is used only once. It can never be stale. That seems a waste, but it does simplify the design.

The third kind of `Var` is a `FunctionVar`. It has no value, just the `FnDef`. The problem here is that functions are global. You can't just put them in the `Vars` list because the variable search starts at `VarSearchStart` -- functions could not call functions. The `FunctionVar` still goes in `Vars`, but its `FnDef` also goes into a global `FunctionTable`, just as in old code. However, the `FunctionTable` is now a `HashMap`. You get the name -- a token ending in '(' -- from the command line, then use that to directly look up the `FnDef`. `GW_lib` is hundreds of functions, so you should see some improvement in speed just because of the faster function look-up. (Look at `isUserFunction()`). 

The change in variable management affects how `doUserFunction()` handles function parameters, especially the pass-by-reference kind. Again, the structure of the code has not changed, only the details. The `FunctionParameter` list `parms` is built the same way as before, but now it holds the entire `Var` for each parameter.

There are some corollary changes. The Dialogs can't use a `Bundle` to pass parameters because I didn't make the `Var` class serializable. I added a new class `DialogArgs` for that purpose. the code for `GR.Set.Pixels' has to have the whole array now, not just pointers into the `Vals` list. And I don't have all of the `Debug` commands caught up with the new system. (I haven't debugged `Debug`.)

There were a few unnecessary changes, too. I moved the array index calculation from `GetArrayValue()` to the `ArrayDef`. With the new `FunctionTable` (now `mFunctionTable`) I could simplify `ESE()` -- it doesn't reparse the same characters as many times, so it's shorter and should be faster. I'd like to do the same kind of thing for `ENE()`.

This is a huge change, but it could have been bigger. I held back almost everything I could. On top of that, we're going to find this new structure will let us do things in the future that we couldn't do be before. For example, this code does not fix the problem of `UNDIM` in an ISR, but now it will be fairly simple to create separate variable lists for different contexts.

But first we have to get this code tested and fully functional again.

**Comment by mougino** [comment]
Thanks for that! :) pretty dense but brilliant! I'll wait for the full `Var.java` and the modified `Run.java` to read and understand better how this works.

I suppose you also had global variables in mind when you re-architected all of that? (but that's a different topic, don't answer that now)

I'm impatient to test it against the GW lib and my RPG to see noticeable speed differences :)

**Comment by jMarcS** [comment]
I had no trouble backing out the binary search for v01.89, but I forgot again to put the Issue number in my commit comment.

The Save and Run bug was pretty simple, too, but I had lots of trouble getting the Editor Load/Save/Delete paths working better. I hope users find the new gimmick useful. But that's off-topic.

I pushed both commits a few minutes ago.

Nicolas, as you suggested, I will release v01.89 and then immediately put the new variable code in v01.89.01. If you would prefer not to wait that long, I can put the new code on GitHub in a branch, or send you the (two) files directly. There is no overlap with the Load/Save/Delete changes.

It has now been over two weeks since I was in the forum. I am terrified of what I will find there!

**Comment by mougino** [comment]
I can wait no worries.

Yes people have started asking where you are ;)
Good luck catching up!

**Comment by humpity** [comment]
> sVarNames = VarNames.size();
   > VarSearchStart = sVarNames;
   > for (FunctionParameter parm : parms) {
   >     int index = newVarIndex(parm.name(), VarNames, sVarNames);

>This fails in any program that uses a function parameter name that matches the name of a variable that already exists in the caller's name space. It's amazing that so many tests worked!

But newVarIndex doesn't use VarSearchStart ? It gets sublistStart from sVarNames. Where does VarSearchStart matter here?

**Comment by jMarcS** [comment]
Yiii! You're right! So my question is: why did changing `VarSearchStart` fix anything? Stay tuned...

**Comment by jMarcS** [comment]
D'oh! It's obvious. In the new code, `doUserFunction()` doesn't call `newVarIndex()`. It looks like this:

    sVarNames = VarNames.size();
    VarSearchStart = sVarNames;
    for (Var.FunctionParameter parm : parms) {
        createNewVar(parm.var());
    }

And `createNewVar()` does use `VarSearchStart`.

So it was not an error in the original `binarySearch()`-based code. It was an error in the changes I made based on the new `Var` architecture.

That explains why it fixed my new code but it did not fix whatever is still wrong with the original var search speed-up code.

I went through the Bug Report forum today and found a post from @evolbug that gave me a simple test case for some problem in the original speed-up code. After v01.89 is released, I plan to take a look at that case.

**Comment by olimaticer** [comment]
I hope, I'm on the right track.
I have made a small drawing to make Marc's briefing visible.
http://1drv.ms/1mANFSG

**Comment by jMarcS** [comment]
That is a good idea to make a drawing of the classes. After I get v01.89 out, I'll try to make one, too.

There are two parallel lists: `VarNames` and `Vars`. `VarNames` is redundant; it is there only for fast binary searches. Anything that has a name has a type. Every `Var` object in the `Vars` list knows its own name and its own type.

I have not made `Var` subclasses for bundles, lists, and stacks, because they do not have names.

In your drawing, I think your box `Var types` should be in blue, too. Is that correct?

You want a list of types to parallel the list of var names and the list of vars. Each `TypeVar` holds type information about another variable.

But the other variable already knows its own type. There is no need for a list of types.

In fact, I have just gone to a lot of trouble to eliminate all those other lists that were needed because a variable did not know its own type. So I do not understand why you want to put it back.

In your previous post, you used a structure like:

    s | Double | Free | 10.3

That is what I just built. There is a `ScalarVar` whose name is "s". It holds a `NumVal`, which holds a `double` (not a `Double`) with the value 10.3. The `Free` field is irrelevant because all variable types are fixed by their names - the name is "s", not "s$" or "s[" or "s$[" or "s(" or "s$(", so it is a numeric scalar variable.

So I do not understand what you are suggesting. Can you clarify?

**Comment by olimaticer** [comment]
Marc, you wrote:

>Var and Val and their subclasses, along with ArrayDef, FnDef, FunctionParameter, and the `Type enum` are all in the file Var.java. It's right around 500 lines, although Run.java is only about 300 lines shorter. 

In Run.Java ****VarType enum**** -->  private enum VarType is a function. So I decoded in my mind "Type enum" is a new Array with a common index slider. I changed the name to VarType. 
Now I know I'm wrong. You mean with (,[,$,$(,$[ or nothing is the var type declared. These characters are also part of var name and in this way also content of Var.java.

Sooner or later the time will come, we will discuss the use of BigInteger, BigDecimal and Integer, too.
Is adding the “!”  for integer values like s! or s!! for BigInterger the solution?
I don't think so.
In future we need a Type class, because in my opinion we will get the TYPE command for more numeric and complexer data types like,
TYPE s AS BigInteger. 
In this case the (,[,$,$(,$[ characters in var names could be obsolet.

Type class examples:
Integer
BigInteger
BigDecimal
Func
FuncOverWrite 'Main function overwrites include function
Del  'Var record could be deleted in idle mode.

To be reckless. 
The table and Type class could contain commands, too.
The preprocessor detects the command and the right record points directly to the branch address.
```    
BASIC! Code:
R =  cos(2*PI) + 45 +5%Example
Object Code may be like this:
LET#R#=#~0
cos#(#~1          OR            executeMF_COS#(#~1 
2#*#PI#~2
)#~1
+#50#~0
```

**Comment by olimaticer** [comment]
Or we ignore gracefully Integer, BigInteger and BigDecimal type classes with a new function:
```
BigDecimalADD$ (a$, b$, D or D$)
BigDecimalSUB$ (a$, b$, D or D$)
BigDecimalMUL$ (a$, b$, D or D$)
BigDecimalDIV$ (a$, b$, D or D$)

D → Digits after point
D$ → Format string
```

/Gregor

**Comment by jMarcS** [comment]
I have committed the new code and pushed it to GitHub. Woohoo!

    marcs@Copperwall:~/basic/ws/BasicWIP4$ git commit
    [master 43eacd3] Reinstate binarySearch. Near total rewrite of variable management.
     3 files changed, 1538 insertions(+), 1313 deletions(-)
     create mode 100644 src/com/rfo/basic/Var.java
    marcs@Copperwall:~/basic/ws/BasicWIP4$

Nicolas, I know you've been waiting for me to get `binarySearch()` back in, and I apologize for taking so long.

Oddly enough, the thing I am most excited about in all this is the rework of arrays. This morning I rewrote the new `Array.fill` command using one of the Java `Arrays.fill()` methods. Nice.

I would like to make this v01.89.01 so we can get testing on real programs by forum users. I'll post about that in the forum if I can find time. Do we need to let v01.89 settle a few days first?

**Comment by mougino** [comment]
Woohoo indeed :)
Both versions can coexist, one official and one for testers. 
Release it Marc! Make it free like an eaaagle :D

**Comment by jMarcS** [comment]
And I did! I released v01.89.01 on December 31. I forgot to put the issue number in the commit (again). The bug reports are coming in. The worst is brochi's array indexing bug: http://rfobasic.freeforums.org/post26295.html#p26295

Nicolas, I hope that your graphics bugs are actually instances of brochi's array indexing bug. There was one real dumb mistake in `gr.text.skew`, but that's all I've found in graphics so far.

The array indexing bug is like this:
`a[i] = a[j] + x`

The array `a[` gets an `ArrayVar` with an `ArrayDef` and an `ArrayVal`.
The index `i` and the value of `a[i]` are cached in the `ArrayVal`.
The index `j` and the value of `a[j]` are cached, too -- in the *same* `ArrayVal`. Oops.
Add in `x` and assign the value to the `ArrayVal`, which writes through to `a[j]`, not `a[i]`.

**Comment by mougino** [comment]
I looked at my code but I don't think the problem is due to an array... This is really a `GR.TEXT.ALIGN` issue.
I start at the very beginning to center text, then for just a few elements i left-align, and immediately center again until the end of the program. 
The second centering is not taken into account in v01.89.01...

**Comment by jMarcS** [comment]
Referring to the failing code snippet in the previous post, the fix is to keep separate `Val` objects for `a[i]` and `a[j]`. We don't really need separate copies unless one is for read and the other is for write. Unfortunately, it's hard to distinguish the cases at the time the caching is done.

My change is to always create a new `ArrayVal` every time the array is indexed. That means reading an array in a loop creates a new `ArrayVal` for every item, instead of re-using the same one repeatedly. I have not measured the effect. I hope it's not too bad, because I don't see another choice.

I tried to make the cost a little less by simplifying the `ArrayVal`. It doesn't cache a value any more -- in fact, it no longer has a member `NumVal` or `StrVal`. It caches only the index.

I also reworked `ArrayDef` so it is an abstract parent of two subclasses, `NumArrayDef` and `StrArrayDef`, with a static factory method to create one or the other as needed. Probably not really necessary, and it does cause a two-line change in `Run.BuildBasicArray()`.

Most of these `Var` and `Val` objects have to be either numeric or string. Sometimes I have one class that checks the type explicitly to decide what to do. Other times I have an abstract class with a subclass for each type -- there's still a type-based decision, but it's made polymorphically by the runtime. I suppose I ought to pick one and use it consistently.

**Comment by jMarcS** [comment]
Okay, Nicolas. I'll look again before cutting v01.89.02. It may be that the problem is not in the `Align` command but in the way text attributes are propagated with the Current Paint. That would probably affect other text attributes, too.

If you want to test the changes I've made already, you can get them from GitHub. I pushed fixes for the `GR.Text.Skew` bug and the array assignment bug. That was the 496th commit -- coming up on 500! (I left out the issue number again but this time it was because the commit comment was already long.)

**Comment by mougino** [comment]
Ok I'll compile up-to-date GitHub source tomorrow morning and report. I was thinking too that this may be a Paint problem. If my `GR.ALIGN` issue is not fixed I'll provide a code snippet you can use for test.

**Comment by mougino** [comment]
After studying my code a little more... 
I'm thinking maybe the problem is that `GR.BITMAP.DRAWINTO.START` newly resets the current Paint??

**Comment by mougino** [comment]
I'm here to report: with GitHub repo up to now I still have the graphic text alignment problem.

I wasn't able to replicate with smaller code, especially I wrote a snippet around `GR.BITMAP.DRAWINTO.START` but it doesn't create (reset) a new Paint as I initially thought...

I'll continue my investigations.

**Comment by jMarcS** [comment]
I found another bug in the new `Var` management. This is not related to the `binarySearch()` speed-up. It's a problem in handling variable names in user-defined function parameters.

The f19_towers_of_hanoi.bas sample program fails. It's a recursive algorithm. The `fn.def` looks like this:
`FN.DEF hanoi(disk,dest,source,other, peg[],  parms[])`
the first call is
`n = hanoi(disk_count, 3, 1, 2, peg[], parms[])`
and it recurses like this:
`n = hanoi(disk-1,other,source,dest, peg[], parms[])`

The first time the `hanoi` function is called, its four scalar parameters are (4,3,1,2). The first recursive call should have parameters (3,2,1,3), but in fact it has (3,2,1,2).

The `fn.def` creates a list of `Var` objects, one for each parameter.

`doUserFunction` can use these variables to hold values until it can move `VarSearchStart`. Then it adds references to those variables to the variable list, starting at the new `VarSearchStart`.

For the recursive call, `doUserFunction` does the same thing again: it uses the list of `Var` objects from `fn.def` to hold temporary values. The problem is that they're the same `Var` objects. It's a little mind-bending, but you can work through it CPU-style, one step at a time:

The `fn.def` list has four `Var` objects. Let's say they have Android object IDs 1,2,3, and 4. `Var` 1 has the name string `disk`, 2 is `dest`, 3 is `source`, and 4 is `other`.

The first `doUserFunction` call puts the value 4 in `Var` 1, 3 in `Var` 2, 1 in `Var` 3, and 2 in `Var` 4. Then it puts pointers to those four objects in the variable list.

The second `doUserFunction` call, one parameter at a time:
`Var` 1: evaluate the expression `disk - 1`, that is, get the caller's variable `disk` (that's `Var` 1) and subtract 1. Assign the result (3) to the first `fn.def` variable -- still `Var` 1. Already a problem, but we get away with this one.
`Var` 2: evaluate the expression `other`, that is, get the caller's variable `other` (that's `Var` 4). Assign its value (2) to the second `fn.def` variable -- that's `Var` 2, which is also the caller's variable `dest`. Yes, we just changed the value of a caller's variable.
`Var` 3: evaluate caller's `source` (`Var` 3) and assign its value (1) to the third `fn.def` variable (`Var` 3). We get away with this one, too.
`Var` 4: evaluate caller's `dest`. That's `Var` 2, which should be 3, but it got changed to 2 when evaluating the second parameter. Assign the value (2) to the fourth `fn.def` variable (`Var` 4).

It's hard to believe this doesn't destroy programs all over the place. Maybe there are not many people using v01.89.01 or v01.89.02? But v01.89.02 runs Nicolas' very large "test" program! Programming never ceases to amaze me.

This bug was probably harder to describe than it will be to fix.

**Comment by jMarcS** [comment]
On the off-chance you're looking at the changes, you may see that I changed the customized `clone()` methods in Var.java to `copy()`. I think it was bad form to mess with the Java definition of "clone".

in `doUserFunction()`, the bugfix is to make a copy of non-global ScalarVar parameters, create a new Val (because the copy is shallow, it uses the same Val as the original), put the new Val in the copy, and puts the copy back in the `FunctionParameter` object.

- The Good: removes one `isNumeric()` test. And it fixes the bug.
- The Bad: making a copy adds code and uses memory. Call it a necessary evil.
- The Ugly: we make the copy and stuff it into `parm` for every non-global ScalarVar parameter of every function call, whether we need to or not. I don't know how to tell if we need to. The original `Var` in each such `parm` is already a copy, and that copy is copied without ever being used.

I'm sure there's a better way, but I'm just not seeing it, so most function calls just got a little slower. But the bug is dead.

**Comment by jMarcS** [comment]
Released in v01.90, 2016/03/31.
